File: examples/python/algorithms_demo.py
def bubble_sort(arr):
    n = len(arr)
    for i in range(n - 1):
        for j in range(n - i - 1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
    return arr

def selection_sort(arr):
    n = len(arr)
    for i in range(n):
        min_idx = i
        for j in range(i + 1, n):
            if arr[j] < arr[min_idx]:
                min_idx = j
        arr[i], arr[min_idx] = arr[min_idx], arr[i]
    return arr

def binary_search(arr, target):
    left, right = 0, len(arr) - 1
    
    while left <= right:
        mid = (left + right) // 2
        
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1
    
    return -1

def linear_search(arr, target):
    for i in range(len(arr)):
        if arr[i] == target:
            return i
    return -1

def find_max(arr):
    if not arr:
        return None
    
    max_val = arr[0]
    for val in arr:
        if val > max_val:
            max_val = val
    return max_val

def find_min(arr):
    if not arr:
        return None
    
    min_val = arr[0]
    for val in arr:
        if val < min_val:
            min_val = val
    return min_val

def calculate_average(arr):
    if not arr:
        return 0
    
    total = sum(arr)
    return total / len(arr)

def main():
    print("Python Algorithms and Data Structures Demo")
    print("=" * 45)
    
    # Test data
    test_array = [64, 34, 25, 12, 22, 11, 90, 5]
    print(f"\nOriginal array: {test_array}")
    
    # Test bubble sort
    bubble_sorted = bubble_sort(test_array.copy())
    print(f"Bubble sorted: {bubble_sorted}")
    
    # Test selection sort
    selection_sorted = selection_sort(test_array.copy())
    print(f"Selection sorted: {selection_sorted}")
    
    # Test search algorithms
    sorted_array = [1, 3, 5, 7, 9, 11, 13, 15, 17, 19]
    target = 7
    
    print(f"\nSearching for {target} in {sorted_array}")
    
    binary_result = binary_search(sorted_array, target)
    linear_result = linear_search(sorted_array, target)
    
    print(f"Binary search result: index {binary_result}")
    print(f"Linear search result: index {linear_result}")
    
    # Test min/max/average
    numbers = [45, 23, 78, 12, 67, 34, 89, 56]
    print(f"\nArray: {numbers}")
    print(f"Maximum: {find_max(numbers)}")
    print(f"Minimum: {find_min(numbers)}")
    print(f"Average: {calculate_average(numbers):.2f}")
    
    # Test with multiple arrays
    test_arrays = [
        [5, 2, 8, 1, 9],
        [100, 50, 75, 25, 80],
        [3, 1, 4, 1, 5, 9, 2, 6]
    ]
    
    print("\nTesting with multiple arrays:")
    for i, arr in enumerate(test_arrays):
        original = arr.copy()
        sorted_arr = bubble_sort(arr)
        print(f"Array {i + 1}: {original} -> {sorted_arr}")
    
    print("\nDemo completed successfully!")

if __name__ == "__main__":
    main()

File: examples/go/algorithms_demo.go
package main

import "fmt"

func bubbleSort(arr []int) {
    n := len(arr)
    for i := 0; i < n-1; i++ {
        for j := 0; j < n-i-1; j++ {
            if arr[j] > arr[j+1] {
                arr[j], arr[j+1] = arr[j+1], arr[j]
            }
        }
    }
}

func quickSort(arr []int, low, high int) {
    if low < high {
        pi := partition(arr, low, high)
        quickSort(arr, low, pi-1)
        quickSort(arr, pi+1, high)
    }
}

func partition(arr []int, low, high int) int {
    pivot := arr[high]
    i := low - 1
    
    for j := low; j < high; j++ {
        if arr[j] < pivot {
            i++
            arr[i], arr[j] = arr[j], arr[i]
        }
    }
    arr[i+1], arr[high] = arr[high], arr[i+1]
    return i + 1
}

func binarySearch(arr []int, target int) int {
    left, right := 0, len(arr)-1
    
    for left <= right {
        mid := left + (right-left)/2
        
        if arr[mid] == target {
            return mid
        }
        if arr[mid] < target {
            left = mid + 1
        } else {
            right = mid - 1
        }
    }
    return -1
}

func findMax(arr []int) int {
    if len(arr) == 0 {
        return 0
    }
    max := arr[0]
    for _, val := range arr {
        if val > max {
            max = val
        }
    }
    return max
}

func main() {
    fmt.Println("Algorithms and Data Structures Demo")
    
    // Test bubble sort
    arr1 := []int{64, 34, 25, 12, 22, 11, 90}
    fmt.Printf("Original array: %v\n", arr1)
    bubbleSort(arr1)
    fmt.Printf("Bubble sorted: %v\n", arr1)
    
    // Test quick sort
    arr2 := []int{10, 7, 8, 9, 1, 5}
    fmt.Printf("Original array: %v\n", arr2)
    quickSort(arr2, 0, len(arr2)-1)
    fmt.Printf("Quick sorted: %v\n", arr2)
    
    // Test binary search
    sortedArr := []int{2, 3, 4, 10, 40}
    target := 10
    result := binarySearch(sortedArr, target)
    if result != -1 {
        fmt.Printf("Element %d found at index %d\n", target, result)
    } else {
        fmt.Printf("Element %d not found\n", target)
    }
    
    // Test find max
    arr3 := []int{5, 2, 8, 1, 9, 3}
    maxVal := findMax(arr3)
    fmt.Printf("Maximum value in %v is %d\n", arr3, maxVal)
    
    // Test with different arrays
    testArrays := [][]int{
        {5, 2, 8, 1, 9},
        {100, 50, 75, 25},
        {1, 3, 2, 4, 5},
    }
    
    fmt.Println("Testing with multiple arrays:")
    for i, arr := range testArrays {
        original := make([]int, len(arr))
        copy(original, arr)
        
        bubbleSort(arr)
        fmt.Printf("Array %d: %v -> %v\n", i+1, original, arr)
    }
}